crates/key-value-azure/src/lib.rs

```diff
@@ -0,0 +1,166 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use azure_data_cosmos::{
+    prelude::{AuthorizationToken, CollectionClient, CosmosClient, Query},
+    CosmosEntity,
+};
+use futures::StreamExt;
+use serde::{Deserialize, Serialize};
+use spin_core::async_trait;
+use spin_key_value::{log_error, Error, Store, StoreManager};
+use tokio::sync::{Mutex, OnceCell};
+
+pub struct KeyValueAzureCosmos {
+    key: String,
+    account: String,
+    database: String,
+    container: String,
+
+    client: OnceCell<Arc<Mutex<CollectionClient>>>,
+}
+
+impl KeyValueAzureCosmos {
+    pub fn new(key: String, account: String, database: String, container: String) -> Result<Self> {
+        Ok(Self {
+            key,
+            account,
+            database,
+            container,
+            client: OnceCell::new(),
+        })
+    }
+}
+
+#[async_trait]
+impl StoreManager for KeyValueAzureCosmos {
+    async fn get(&self, name: &str) -> Result<Arc<dyn Store>, Error> {
+        let client = self
+            .client
+            .get_or_try_init(|| async {
+                let token =
+                    AuthorizationToken::primary_from_base64(&self.key).map_err(log_error)?;
+
+                let cosmos_client = CosmosClient::new(self.account.clone(), token);
+                let database_client = cosmos_client.database_client(self.database.clone());
+
+                let collection_client = database_client.collection_client(self.container.clone());
+                Ok(Arc::new(Mutex::new(collection_client)))
+            })
+            .await?;
```

Comment:
Is there a reason why we're initializing here instead of in `KeyValueAzureCosmos::new`?